File: 2657/ts/main.ts
function findThePrefixCommonArray(A: number[], B: number[]): number[] {
  const n = A.length;
  const ans: number[] = [];
  const freq: number[] = new Array(n + 1);
  let count = 0;

  for (let i = 0; i < n; i++) {
    const a = A[i], b = B[i];

    freq[a] = (freq[a] || 0) + 1;
    if (freq[a] === 2) {
      count++;
    }

    freq[b] = (freq[b] || 0) + 1;

    if (freq[b] === 2) {
      count++;
    }

    ans.push(count);
  }

  return ans;
}

let A = [1, 3, 2, 4], B = [3, 1, 2, 4];

console.log(findThePrefixCommonArray(A, B));

A = [2, 3, 1], B = [3, 1, 2];

console.log(findThePrefixCommonArray(A, B));

File: 2657/go/main.go
package main

import "fmt"

func findThePrefixCommonArray(A []int, B []int) []int {
	n := len(A)

	count := 0
	ans := []int{}
	freq := make([]int, n+1)

	for i, a := range A {

		b := B[i]
		freq[a]++

		if freq[a] == 2 {
			count++
		}

		freq[b]++
		if freq[b] == 2 {
			count++
		}
		ans = append(ans, count)
	}

	return ans
}

func main() {
	fmt.Println(findThePrefixCommonArray(
		[]int{1, 3, 2, 4},
		[]int{3, 1, 2, 4},
	))

	fmt.Println(findThePrefixCommonArray(
		[]int{2, 3, 1},
		[]int{3, 1, 2},
	))

}
